src/transcription/node/utils.ts

```diff
@@ -128,6 +131,149 @@ const signFile = async (axiosInstance: AxiosInstance, url: string, fileName: str
+ * @param parts - Array of part information objects (each contains etag and partNumber).
+ * @returns Response to the complete upload request.
+ */
+const completeUpload = async (
+  axiosInstance: AxiosInstance,
+  url: string,
+  uploadId: string,
+  parts: { etag: string; partNumber: number }[],
+): Promise<any> => {
+  const completeUrl = `${url}?action=mpu-complete&uploadId=${uploadId}`
+  const response = await axiosInstance.put(completeUrl, JSON.stringify({ parts }), {
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  })
+  return response
+}
+
+/**
+ * Reads a file in chunks and processes each chunk with a provided callback function.
+ * @param filePath - The full path of the file to be read.
+ * @param fileSize - The total size of the file in bytes.
+ * @param maxChunkSizeBytes - The maximum size for each chunk.
+ * @param eachChunk - Async callback that processes each chunk; should return an object containing etag and partNumber.
+ * @returns A promise that resolves to an array of results from processing each chunk.
+ */
+const readFileInChunks = async (
+  filePath: string,
+  fileSize: number,
+  maxChunkSizeBytes: number,
+  eachChunk: (chunkNumber: number, chunk: Buffer) => Promise<{ etag: string; partNumber: number }>,
+): Promise<{ etag: string; partNumber: number }[]> => {
+  let fileHandle: fs.FileHandle
+  try {
+    fileHandle = await fs.open(filePath, 'r')
+  } catch (e: any) {
+    throw new Error(`Error opening file: ${filePath}`)
+  }
+
+  const numChunks = Math.ceil(fileSize / maxChunkSizeBytes)
+  const realChunkSize = Math.ceil(fileSize / numChunks)
+
+  let bytesRead = 0
+  let chunkNumber = 1
+  const allChunks: Promise<{ etag: string; partNumber: number }>[] = []
+
+  while (chunkNumber <= numChunks) {
+    const buffer = Buffer.alloc(realChunkSize)
+    const { bytesRead: bytesJustRead } = await fileHandle.read(buffer, 0, realChunkSize, bytesRead)
+    bytesRead += bytesJustRead
+    allChunks.push(eachChunk(chunkNumber, buffer))
+    chunkNumber++
+  }
+
+  await fileHandle.close()
+  return Promise.all(allChunks)
+}
+
+/**
+ * Orchestrates the multipart file upload process:
+ * 1. Initiates the upload session.
+ * 2. Reads the file in chunks and uploads each part.
+ * 3. Completes the upload once all parts are processed.
+ *
+ * @param axiosInstance - The Axios instance for making HTTP requests.
+ * @param fileName - The name of the file being uploaded.
+ * @param filePath - The path of the file on the local filesystem.
+ * @param fileSize - The total size of the file in bytes.
+ * @param organizationName - The name of the organization (used in URL paths).
+ * @param partSizeBytes - The maximum size in bytes for each file part.
+ */
+const uploadFileInParts = async (
+  axiosInstance: AxiosInstance,
+  fileName: string,
+  filePath: string,
+  fileSize: number,
+  organizationName: string,
+  partSizeBytes: number,
+): Promise<void> => {
```

Comment:
If users have a lot of large transcriptions to upload, this could be really time consuming  especially when they're uploading multiple large files in a row.
It might be worth giving them a way to escape the loop if it starts taking too long.

Looking into standards, I found that `AbortController `is a popular and well-supported way to handle cancellations in long-running async operations.



```suggestion
const uploadFileInParts = async (
  axiosInstance: AxiosInstance,
  fileName: string,
  filePath: string,
  fileSize: number,
  organizationName: string,
  partSizeBytes: number,
  signal?: AbortSignal
): Promise<void> => {
```
Then below do something like:
```
 if (signal?.aborted) {
      throw new Error('Upload cancelled by user')
    }
```

Reply:
We've already used `AbortController` in waitFor. I agree; it's a good idea to give a way to escape the loop if it starts taking too long.
Updated `transcribe` method.